### PR TITLE
Upstream envoyproxy/envoy-wasm.

### PR DESCRIPTION
**Pull request by jplevyak**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: Upstream envoyproxy/envoy-wasm.
Additional Description: This PR represents the proxy-wasm work which has been going on over the last 18 months in the envoyproxy/envoy-wasm repository.  The vast majority of the changes herein are confined to extensions which are optional.  Changes outside of the extensions are the minimal set required in order to support those extensions.
Risk Level: Low (if the extensions are not included).
Testing: Unit tests in envoyproxy/envoy and integration tests in istio/proxy.
Docs Changes: Docs will be provided in subsequent PRs.
Release Notes: Wasm plugins available optionally.
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/4272
[Optional Deprecated:]

### CODE REVIEW

File: api/envoy/extensions/wasm/v3/wasm.proto
Comment (lizan) [inline api/envoy/extensions/wasm/v3/wasm.proto:37]: The original description should be kept?

File: api/envoy/extensions/wasm/v3/wasm.proto
Comment (lizan) [inline api/envoy/extensions/wasm/v3/wasm.proto:67]: Why this changed from last API upstram?

File: api/wasm/README.md
Comment (lizan) [inline api/wasm/README.md:1]: Let's wait for envoyproxy/envoy-wasm#516 (which is ready now) and remove this directory.

File: bazel/BUILD
Comment (lizan) [inline bazel/BUILD:251]: we decided not upstream WAVM meanwhile?

File: api/envoy/extensions/wasm/v3/wasm.proto
Comment (jplevyak) [reply api/envoy/extensions/wasm/v3/wasm.proto:67]: This change was never pulled into envoy-wasm and will require a change in istio.  Given that it isn't implemented in Envoy and is not going to be implemented before v4 I didn't think there was any value in causing code churn at this point for a feature which doesn't exist.  We can have a breaking change in v4 when we undoubtedly will have other breaking changes rather than introducing another one unnecessarily here.

File: api/wasm/README.md
Comment (jplevyak) [reply api/wasm/README.md:1]: I will merge that in and ping this thread.

File: bazel/BUILD
Comment (jplevyak) [reply bazel/BUILD:251]: The performance of WAVM is much higher than V8, consequently the istio team would like to have WAVM as an option and we felt that if istio would want it others would as well.

File: api/envoy/extensions/wasm/v3/wasm.proto
Comment (jplevyak) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Done.

File: bazel/BUILD
Comment (jplevyak) [reply bazel/BUILD:251]: We discussed this with Matt at our last meeting  (you might have come in after that) and he was onboard with including WAVM as an option.

File: api/envoy/extensions/wasm/v3/wasm.proto
Comment (lizan) [reply api/envoy/extensions/wasm/v3/wasm.proto:70]: SGTM.

File: bazel/BUILD
Comment (lizan) [reply bazel/BUILD:251]: Sounds good then. Do we need this config_settings though? Isn't VM implementations are registered via extensions?

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:531]: Latest is 1.40.1
```
    emscripten_toolchain = dict(
        sha256 = "81f620f17d81c18ab84c69ddfcabe813b2bf4250d69729b961ad57101a16c218",
        patch_cmds = [
            "./emsdk install 1.40.1-upstream",
            "./emsdk activate --embedded 1.40.1-upstream",
        ],
        strip_prefix = "emsdk-1.40.1",
        urls = ["https://github.com/emscripten-core/emsdk/archive/1.40.1.tar.gz"],
        use_category = ["build"],
    ),
```

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:518]: As we aren't using release please add a date comment under `strip_prefix` in the format of `# YYYY-MM-DD`

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:511]: As we aren't using release please add a date comment under `strip_prefix` in the format of `# YYYY-MM-DD`

File: bazel/repositories.bzl
Comment (moderation) [inline bazel/repositories.bzl:416]: Use latest 0.5.0
```
        name = "rules_antlr",
        sha256 = "26e6a83c665cf6c1093b628b3a749071322f0f70305d12ede30909695ed85591",
        strip_prefix = "rules_antlr-0.5.0",
        urls = ["https://github.com/marcohu/rules_antlr/archive/0.5.0.tar.gz"],
```

File: bazel/repositories.bzl
Comment (moderation) [inline bazel/repositories.bzl:436]: Can this be bumped to 4.8?
```
        sha256 = "992d52444b81ed75e52ea62f9f38ecb7652d5ce2a2130af143912b3042a6d77e",
        strip_prefix = "antlr4-4.8",
        urls = ["https://github.com/antlr/antlr4/archive/4.8.tar.gz"],
```
In order to use this `bazel_dependency_imports.bzl` needs to change
```diff
diff --git a/bazel/dependency_imports.bzl b/bazel/dependency_imports.bzl
index a0a0110f4..9c826aba8 100644
--- a/bazel/dependency_imports.bzl
+++ b/bazel/dependency_imports.bzl
@@ -7,10 +7,10 @@ load("@build_bazel_rules_apple//apple:repositories.bzl", "apple_rules_dependenci
 load("@upb//bazel:repository_defs.bzl", upb_bazel_version_repository = "bazel_version_repository")
 load("@config_validation_pip3//:requirements.bzl", config_validation_pip_install = "pip_install")
 load("@protodoc_pip3//:requirements.bzl", protodoc_pip_install = "pip_install")
-load("@rules_antlr//antlr:deps.bzl", "antlr_dependencies")
+load("@rules_antlr//antlr:repositories.bzl", "rules_antlr_dependencies")

 def envoy_dependency_imports(go_version = GO_VERSION):
     rules_foreign_cc_dependencies()
@@ -20,7 +20,7 @@ def envoy_dependency_imports(go_version = GO_VERSION):
     gazelle_dependencies()
     apple_rules_dependencies()
     upb_bazel_version_repository(name = "upb_bazel_version")
-    antlr_dependencies(471)
+    rules_antlr_dependencies("4.8")

     custom_exec_properties(
```

File: bazel/repository_locations.bzl
Comment (jplevyak) [reply bazel/repository_locations.bzl:531]: There is an interaction between the toolchain and the newer versions of emscripten that will require some work on the SDK and bazel rules.  Consequently  I am proposing rolling back the version until this can be sorted out.

File: bazel/BUILD
Comment (jplevyak) [reply bazel/BUILD:251]: That is a good point.  Currently it is implemented as a -define wasm=XXXX option.  That was appropriate when we were doing development, but perhaps registering it as an extension would be more appropriate now. I am not sure how that would work though.  Perhaps using a envoy/registry/registry.h and registering a factory for the runtime?

Comment (lizan) [reply bazel/BUILD:251]: yes, feel free to just leave a TODO (with an issue) and tackle it later.

File: bazel/repository_locations.bzl
Comment (moderation) [reply bazel/repository_locations.bzl:531]: Maybe add a TODO?

File: api/wasm/README.md
Comment (jplevyak) [reply api/wasm/README.md:1]: merged

**Comment by jplevyak** [comment]
The Linux-x64 release seems to be a flake.

I don't know that this code has ever been tested on Windows or arm64.  Any suggestions on how to debug these things locally and/or disable the Wasm extensions for these environments?

File: bazel/BUILD
Comment (jplevyak) [reply bazel/BUILD:252]: Done.  With issue https://github.com/envoyproxy/envoy/issues/12574

File: bazel/repository_locations.bzl
Comment (jplevyak) [reply bazel/repository_locations.bzl:532]: done.

File: bazel/repositories.bzl
Comment (jplevyak) [reply bazel/repositories.bzl:420]: Upgrading this fails to compile because "headers" is no longer supported in 0.5.0.  CEL needs to be upgraded first to support 0.5.0.  Added TODO to upgrade this when CEL is upgraded.

File: bazel/repositories.bzl
Comment (jplevyak) [reply bazel/repositories.bzl:440]: Because I can't update antlr_rules I can't bump this to 4.8 since that version is not supported by the current version of antlr_rules.

File: bazel/repository_locations.bzl
Comment (jplevyak) [reply bazel/repository_locations.bzl]: Done.

File: .bazelrc
Comment (lizan) [inline .bazelrc:188]:
```suggestion
build:remote --strategy=Rustc=sandboxed,local
```

This was my oversight in envoy-wasm PR to actually enable Rust works in envoy CI.

**Comment by lizan** [comment]
@jplevyak we can skip arm / windows, just modify build_setup / do_ci script to add bazel flags.

**Comment by htuch** [comment]
@jplevyak how far does this get us to Wasm upstreaming? Is this mission complete? Would be good to document the support status after this PR in the commit message and version history.

File: api/wasm/README.md
Comment (htuch) [inline api/wasm/README.md:1]: The `api/` tree is for xDS APIs, why is there Wasm docs and build files here?

File: source/extensions/access_loggers/wasm/BUILD
Comment (htuch) [inline source/extensions/access_loggers/wasm/BUILD:27]: You probably also want to tag this as alpha.

File: source/extensions/bootstrap/wasm/BUILD
Comment (htuch) [inline source/extensions/bootstrap/wasm/BUILD:19]: Ditto, alpha here too.

File: source/extensions/filters/http/wasm/BUILD
Comment (htuch) [inline source/extensions/filters/http/wasm/BUILD:32]: alpha again

File: source/extensions/filters/network/wasm/BUILD
Comment (htuch) [inline source/extensions/filters/network/wasm/BUILD:31]: also alpha

File: bazel/foreign_cc/llvm.patch
Comment (htuch) [inline bazel/foreign_cc/llvm.patch:9]: Can you upstream this? We want as few patches as possible to improve update velocity.

File: bazel/foreign_cc/BUILD
Comment (htuch) [inline bazel/foreign_cc/BUILD:231]: @lizan thoughts on relationship to https://github.com/envoyproxy/envoy/blob/master/tools/clang_tools/support/BUILD.prebuilt?

**Comment by jplevyak** [comment]
@htuch with this we will be 95% complete.  There are a couple things which are still under development by interns and there is one thing that istio needs, but for all intents and purposes this will be enough for most users.

**Comment by jplevyak** [comment]
> @jplevyak we can skip arm / windows, just modify build_setup / do_ci script to add bazel flags.

Which flags do you mean? I don't think we have a flag to remove Wasm support i.e. we have --define wasm=enabled and --define wsam=wavm, but I don't think we have --define wasm=none.

File: source/extensions/filters/network/wasm/BUILD
Comment (jplevyak) [reply source/extensions/filters/network/wasm/BUILD:31]: What is the syntax?  "unknown alpha" "unknown, alpha" or just "alpha" ?

**Comment by PiotrSikora** [comment]
> > @jplevyak we can skip arm / windows, just modify build_setup / do_ci script to add bazel flags.
> 
> Which flags do you mean? I don't think we have a flag to remove Wasm support i.e. we have --define wasm=enabled and --define wsam=wavm, but I don't think we have --define wasm=none.

We used to have `wasm=disabled` before we enabled V8 by default. It should be in archives if you want to dig it up.

File: source/extensions/filters/network/wasm/BUILD
Comment (htuch) [reply source/extensions/filters/network/wasm/BUILD:31]: See https://github.com/envoyproxy/envoy/blob/da403fa7b67b9331ca1d129a4035950d2e23c752/source/extensions/grpc_credentials/file_based_metadata/BUILD#L19.

File: tools/spelling/spelling_dictionary.txt
Comment (ggreenway) [inline tools/spelling/spelling_dictionary.txt:272]: For at least some of the words in this file, consider putting them in backticks in comments; that will make the spellchecker ignore them.

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [inline source/server/admin/prometheus_stats.cc:198]: Revert. This has already been handled in an alternate way.

**Comment by ggreenway** [comment]
> > @jplevyak we can skip arm / windows, just modify build_setup / do_ci script to add bazel flags.
> 
> Which flags do you mean? I don't think we have a flag to remove Wasm support i.e. we have --define wasm=enabled and --define wsam=wavm, but I don't think we have --define wasm=none.

Please add a flag for `wasm=none`. It's a huge dependency, and should be optional.

File: source/common/stats/thread_local_store.h
Comment (jmarantz) [inline source/common/stats/thread_local_store.h:335]: are there any tests that fail if we don't make this particular change?

scope names are sanitized I think.

File: api/wasm/README.md
Comment (jplevyak) [reply api/wasm/README.md:1]: This is all going to be removed.  Pending https://github.com/envoyproxy/envoy-wasm/pull/613 which I will merge in here and ping this thread.

File: source/extensions/access_loggers/wasm/BUILD
Comment (jplevyak) [reply source/extensions/access_loggers/wasm/BUILD:27]: Done

File: source/extensions/filters/http/wasm/BUILD
Comment (jplevyak) [reply source/extensions/filters/http/wasm/BUILD:32]: Done

File: source/extensions/filters/network/wasm/BUILD
Comment (jplevyak) [reply source/extensions/filters/network/wasm/BUILD:31]: Done

**Comment by jplevyak** [comment]
@lizan I have made v8 support optional.  We now need to add --define wasm=enabled to make it enabled.  Which C builds should it default to being enabled?

**Comment by ggreenway** [comment]
> @lizan I have made v8 support optional. We now need to add --define wasm=enabled to make it enabled. Which C builds should it default to being enabled?

In general, most/all things default to enabled/included. Also, please the settings here: https://github.com/envoyproxy/envoy/blob/master/bazel/README.md#disabling-optional-features

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (htuch) [inline bazel/external/proxy_wasm_cpp_host.BUILD:23]: Wavm has a cmake build system, why can't we build it under rules_foreign_cc rather than this handcrafted blob thing?

Comment (PiotrSikora) [reply bazel/external/proxy_wasm_cpp_host.BUILD:23]: WAVM is built using `rules_foreign_cc`, this here is a `BUILD` file for Proxy-Wasm C++ host implementation, which optionally includes WAVM support.

Comment (htuch) [reply bazel/external/proxy_wasm_cpp_host.BUILD:23]: Is there any possibility here of not checking in this long glob file?

File: bazel/repositories.bzl
Comment (kyessenov) [reply bazel/repositories.bzl:421]: Google hasn't imported new ANTLR, so it might take some time to upgrade Google. Current runtime has been fuzzed extensively, so I'm not too worried about missing some bugs.

Comment (htuch) [reply bazel/repositories.bzl:421]: I don't think this will break Google import, since we only allowlist certain extensions and Wasm is not yet on this. @jplevyak  we should follow-up online on what it takes to import.

Comment (kyessenov) [reply bazel/repositories.bzl:421]: I mean cel-cpp is a direct export from google3, so it is linked against internal google3 patched ANTLR which is not based on 0.5.0. We can import 0.5.0 into google3 and then link cel-cpp against that runtime. I am not sure what it takes to do a security review of a parser runtime though. When I looked briefly into that, it looked like ANTLR was lenient with breaking changes in a minor version.

**Comment by PiotrSikora** [comment]
> In general, most/all things default to enabled/included. Also, please the settings here: https://github.com/envoyproxy/envoy/blob/master/bazel/README.md#disabling-optional-features

So what's the default that we want to implement? V8 enabled by default, WAVM (only?) enabled with `--define wasm=wavm`, no Wasm with `--define wasm=disabled`?

**Comment by ggreenway** [comment]
> > In general, most/all things default to enabled/included. Also, please the settings here: https://github.com/envoyproxy/envoy/blob/master/bazel/README.md#disabling-optional-features
> 
> So what's the default that we want to implement? V8 enabled by default, WAVM (only?) enabled with `--define wasm=wavm`, no Wasm with `--define wasm=disabled`? Do we also want both enabled with `--define wasm=enabled`?

I don't know the tradeoffs or considerations for V8 vs WAVM. Someone else can chime on that.

My request is just that there be a way to not compile either one in (which it looks like is now provided), and that we try to test as many variations as possible in CI.

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (jplevyak) [reply bazel/external/proxy_wasm_cpp_host.BUILD:23]: Do you want all the files called out separately for each case?  It will be longer, but perhaps that is OK?

File: tools/spelling/spelling_dictionary.txt
Comment (jplevyak) [reply tools/spelling/spelling_dictionary.txt]: I removed some PTAL

File: source/server/admin/prometheus_stats.cc
Comment (jplevyak) [reply source/server/admin/prometheus_stats.cc]: done.

File: bazel/repositories.bzl
Comment (htuch) [reply bazel/repositories.bzl:421]: Right; are we just talking about BUILD files and rules_antlr here? If so, being behind the last release version isn't a security worry.

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (htuch) [reply bazel/external/proxy_wasm_cpp_host.BUILD:23]: It looks like you folks are contributors to https://github.com/proxy-wasm/proxy-wasm-cpp-host. Can you add some Bazel BUILD files there?

File: bazel/foreign_cc/llvm.patch
Comment (jplevyak) [reply bazel/foreign_cc/llvm.patch:9]: https://github.com/envoyproxy/envoy/issues/12634

Comment (htuch) [reply bazel/foreign_cc/llvm.patch:9]: Same goes for the other patches as well :) We really don't like to accumulate patches; they are easy to write and unblock folks adding new deps, but become a nightmare for maintenance, as it involves costly labor to keep them in sync as we bump version and it impacts our security upgrade velocity as a result.

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (lizan) [inline bazel/external/proxy_wasm_cpp_host.BUILD:7]: This shouldn't leaked here. You should be able to just do select in `deps` where these library target are referenced. The BUILD file itself can live in proxy_wasm_cpp_host repo.

File: bazel/foreign_cc/BUILD
Comment (lizan) [inline bazel/foreign_cc/BUILD:361]: revert

File: ci/do_ci.sh
Comment (lizan) [inline ci/do_ci.sh:267]: seems conflicted?

**Comment by jplevyak** [comment]
@lizan both of the failures seem to be CI related.  Can you confirm?

**Comment by lizan** [comment]
@jplevyak gcc CI is, I'll working on that in another PR. Coverage failure are real:
```
Code coverage 95.9 is lower than limit of 96.5
```
Check the coverage report: https://storage.googleapis.com/envoy-pr/12546/coverage/index.html

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (jplevyak) [reply bazel/external/proxy_wasm_cpp_host.BUILD:7]: I can have additional libraries to allow WAVM and V8 to be included separately in either external/proxy-wasm-cpp-host.BUILD or in the BUILD file in the external repository, but I am having problems with zlib:

The zlib library in envoy is strange, and if I include zlib in proxy-wasm-cpp-host and depend on it via @zlib in the deps, then envoy fails because it wants to see //external:zlib which proxy-wasm-cpp-host/BUILD doesn't like.

The envoy zlib build uses a filegroup then a net_zlib target then maps via anative.bind to zlib from @envoy/baze/foreign_cc::zlib, so there are at least 4 names involved: zlib, net_lib, @envoy/baze/foreign_cc::zlib, and //external:zlib  and it isn't clear to me which one I should be using to get the BUILD file working in envoy and in the external repo.  I have tried a couple combinations to no avail.  Any suggestion?

**Comment by jplevyak** [comment]
@lizan I'll move to working on coverage.  Thanx.

**Comment by jplevyak** [comment]
https://github.com/envoyproxy/envoy/issues/13212 is blocking

**Comment by jplevyak** [comment]
https://github.com/envoyproxy/envoy/issues/13180 is blocking

**Comment by jplevyak** [comment]
ci/circleci: all tests is passing and yet it is marked as failing, not sure why, but it is a CI issue, likely disk space
ci/circleci: asan is failing because it is being killed by circleci. again not sure why but it seems to be a resources problem.
envoy-presubmit is failing because it is running out of disk space on azure
envoy-presubmit (Linux-x64 asan) is failing because of https://github.com/envoyproxy/envoy/issues/13180

**Comment by jplevyak** [comment]
https://dev.azure.com/cncf/envoy/_build/results?buildId=52413&view=logs&j=a5e52b91-c83f-5429-4a68-c246fc63a4f7&t=22325323-cbb2-580d-876a-9766e4a9c105&l=13

Mac OS is failing to publish results.  The tests themselves pass.

**Comment by jplevyak** [comment]
Only issue now is the MacOS CI

**Comment by mattklein123** [comment]
OSX is messed up right now. We can force merge assuming this is ready to go and passes otherwise. Are there any remaining issues? cc @lizan

**Comment by mattklein123** [comment]
@lizan should be back from vacation and will get this merged. Assigning over to him.

**Comment by PiotrSikora** [comment]
@mattklein123 @htuch all tests pass now, and this PR is synced with the current `master`, so it should be good to go now.

**Comment by mattklein123** [comment]
@PiotrSikora DCO is now broken. Can you just squash it all into a single commit and force push?

**Comment by mattklein123** [comment]
Note that we are going to merge this without docs but we need to do a doc pass before we do any major PR about this feature.
Ref: https://github.com/envoyproxy/envoy/issues/13426 https://github.com/envoyproxy/envoy/issues/9465 https://github.com/envoyproxy/envoy/issues/9462

cc @louiscryan

File: CODEOWNERS
Comment (lizan) [inline CODEOWNERS:84]: @PiotrSikora John is no longer working on this?

Comment (PiotrSikora) [reply CODEOWNERS:84]: No, I'll update `CODEOWNERS`.

**Review by lizan — APPROVED** [review]

**Review by mattklein123 — APPROVED** [review]

**Comment by duderino** [comment]
huzzah!

**Comment by gbrail** [comment]
Cool!

**Pull request merged**
